synapse/util/metrics.py

```diff
@@ -220,21 +220,21 @@ class DynamicCollectorRegistry(CollectorRegistry):
 
     def __init__(self) -> None:
         super().__init__()
-        self._pre_update_hooks: List[Callable[[], None]] = []
+        self._pre_update_hooks: Dict[str, Callable[[], None]] = {}
 
     def collect(self) -> Generator[Metric, None, None]:
         """
         Collects metrics, calling pre-update hooks first.
         """
 
-        for pre_update_hook in self._pre_update_hooks:
+        for pre_update_hook in self._pre_update_hooks.values():
             pre_update_hook()
 
         yield from super().collect()
 
-    def register_hook(self, hook: Callable[[], None]) -> None:
+    def register_hook(self, metric_name: str, hook: Callable[[], None]) -> None:
         """
         Registers a hook that is called before metric collection.
         """
 
-        self._pre_update_hooks.append(hook)
+        self._pre_update_hooks[metric_name] = hook
```

Comment:
Should we be doing anything if we're overriding one? Do we need to "unregister" the old one somehow?

Overall this seems to rollback to the previous behavior though so is likely OK.

Reply:
It will be 'unregistered' by being overwritten in the dict; the hook isn't tracked anywhere else.

Reply:
Ah-ha! I thought it somehow got tracked internally from the tracing bits, but doesn't look like it. 👍 Thank you for the explanation!